### PR TITLE
Add generic parameter for an allocator to collection types and box-like structures.

### PR DESCRIPTION
**Pull request by TimDiekmann**
This adds a generic parameter `A: AllocRef` to `Box`. This pull request mainly was made to run crater to test compatibility when adding a defaulted generic parameter to one of the most used structs. It's not yet possible to construct a `Box` with a generic allocator.

cc rust-lang/wg-allocators#1

r? @Amanieu

### CODE REVIEW

**Comment by TimDiekmann** [comment]
How the Try build can succeed, but the PR build fails? (locally, `x.py test` also passed)

**Comment by TimDiekmann** [comment]
Rebased onto master and fixed 32bit output (Try-build only checks 64bit)

**Comment by Amanieu** [comment]
@craterbot check

**Comment by Amanieu** [comment]
Actually, I think we should go ahead and add allocator parameters to `Vec`, `VecDeque`, `LinkedList`, `BinaryHeap`, `BTreeMap`, `BTreeSet`, `HashMap` and `HashSet`. We will need to add allocator support to all of those anyways, so we might as well do a single crater run for all of them.

**Comment by Mark-Simulacrum** [comment]
Historically we've also held off on doing so because it was a pretty sizeable regression on perf, IIRC -- maybe we should check; @rust-timer queue a6b9054912d7518df0cd0c0050d5c5b1a51b9b14

**Comment by Mark-Simulacrum** [comment]
er I meant @rust-timer build a6b9054

**Comment by TimDiekmann** [comment]
> Actually, I think we should go ahead and add allocator parameters to `Vec`, `VecDeque`, `LinkedList`, `BinaryHeap`, `BTreeMap`, `BTreeSet`, `HashMap` and `HashSet`. We will need to add allocator support to all of those anyways, so we might as well do a single crater run for all of them.

One crater run seems good, but for performance measurement I don't think it's the best option. As the timer is queued, I'll add a parameter to the other types. Also, does hashbrown currently support any type of allocator? I'd leave that out for now.

**Comment by TimDiekmann** [comment]
I noticed that most of the collections uses methods like `VecDeque::as_mut_slices` in `Drop`. Should I add those methods for private usage with an allocator, or just change `impl VecDeque<T>` to `VecDeque<T, A>` for those functions?

**Comment by Amanieu** [comment]
`as_mut_slices` doesn't depend on the allocator so it should be safe to make it allocator-generic.

**Comment by TimDiekmann** [comment]
I added a parameter to `BinaryHeap`, `BTreeMap`, `BTreeSet`, `LinkedList`, `VecDeque`, `Rc`, `Arc`, `Vec`, `String`, `HashMap`, and `HashSet`.

**Comment by TimDiekmann** [comment]
Rebased onto master and fixed the `LinkedList` error (Sideeffect of #71878)

**Comment by Amanieu** [comment]
On review, I don't think this is sufficient for a crater run. We also want all the `impl` blocks to switch from `Box<T>` to `Box<T, A>` so that methods are available for all allocators. Only implementing them for `Box<T>` might give us false confidence since type inference will default `A` to `Global` since that is the only allowed value for the methods used.

**Comment by TimDiekmann** [comment]
I adjusted the `Box` implementation and running a test locally right now. The changes got pretty big, should I also implement that for all other collections?

**Comment by bluss** [comment]
Maybe it makes sense to try only for Box. Are there any new features that help type inference here? If it's just a default type parameter on the type, we know that the fallout - non-compiling code - in terms of type inference is going to be pretty large, right?

**Comment by TimDiekmann** [comment]
> Maybe it makes sense to try only for Box. Are there any new features that help type inference here? If it's just a default type parameter on the type, we know that the fallout - non-compiling code - in terms of type inference is going to be pretty large, right?
> 
> In particular, any API that is like `Type::new`, and it gets more generic than before (undetermined allocator parameter), will be a trouble spot - `Box::new` for example.

Only the default parameter on the struct is used. My previous tests has shown, that anything returning `Self` is a problem if `A` cannot be inferred easily, basically all constructors. Trait implementations like `Default` should work just fine. I'm pretty sure that this can be fixed, as calling `Box::new(10)` could be inferred by the compiler to `Box::<i32>::new(10)` which fallbacks to `Box::<i32, Global>::new(10)`. Thus I havn't changed the `impl`-block for `new`, `new_uninit` etc., but added `new_in(A)`. At a later point we could add `impl Box<T, A: Default> { fn new() }`

With the latest commit I got a weird error, is there anything I am missing?

```
# ./x.py test --bless --stage 1
[...]
Testing alloc stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
   Compiling getrandom v0.1.14
   Compiling ppv-lite86 v0.2.6
   Compiling c2-chacha v0.2.3
   Compiling rand_core v0.5.1
   Compiling rand_xorshift v0.2.0
   Compiling rand_chacha v0.2.1
   Compiling rand v0.7.3
   Compiling alloc v0.0.0 (/home/tim/Code/rust/src/liballoc)
error[E0308]: mismatched types
   --> src/liballoc/vec.rs:685:13
    |
679 |     pub fn into_boxed_slice(mut self) -> Box<[T]> {
    |                                          -------- expected `std::boxed::Box<[T]>` because of return type
...
685 |             buf.into_box(len).assume_init()
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `std::alloc::Global`, found struct `alloc::Global`
    |
    = note: expected struct `std::boxed::Box<_, std::alloc::Global>`
               found struct `std::boxed::Box<_, alloc::Global>`
```

`alloc::Global`? Isn't there a crate missing? And isn't `alloc::Global` and `std::alloc::Global` exactly the same struct?

**Comment by TimDiekmann** [comment]
This corresponds to this CI failure: https://github.com/rust-lang/rust/pull/71873/checks?check_run_id=648683135#step:24:1716

File: src/liballoc/boxed.rs [hunk 1]
```diff
@@ -437,13 +587,18 @@ impl<T: ?Sized> Box<T> {
...
-        // mutable reference. Turning *that* to a raw pointer behaves correctly.
-        Box::leak(b) as *mut T
+    pub fn into_raw(b: Self) -> *mut T {
+        Self::into_raw_with_alloc(b).0
+    }
+
+    /// Behaves like [`into_raw`] but also returns the allocator
+    ///
+    /// [`into_raw`]: #method.into_raw
+    #[stable(feature = "box_raw", since = "1.4.0")]
```
Comment (Amanieu) [inline src/liballoc/boxed.rs:597]: This should be unstable under `allocator_api`.

File: src/liballoc/boxed.rs [hunk 2]
```diff
@@ -481,13 +636,19 @@ impl<T: ?Sized> Box<T> {
         reason = "use `Box::leak(b).into()` or `NonNull::from(Box::leak(b))` instead"
     )]
     #[inline]
-    pub fn into_raw_non_null(b: Box<T>) -> NonNull<T> {
-        // Box is recognized as a "unique pointer" by Stacked Borrows, but internally it is a
-        // raw pointer for the type system. Turning it directly into a raw pointer would not be
-        // recognized as "releasing" the unique pointer to permit aliased raw accesses,
-        // so all raw pointer methods go through `leak` which creates a (unique)
-        // mutable reference. Turning *that* to a raw pointer behaves correctly.
```
Comment (Amanieu) [inline src/liballoc/boxed.rs:489]: It would be nice not to lose this comment.

**Comment by Amanieu** [comment]
Let's do a crater run with just `Box` for now. However we will eventually want to do this for all collection types.

**Comment by TimDiekmann** [comment]
Adding `Vec` was pretty straight forward once the `Box` compiled fine. I think adding collections which relies on `Vec` or `RawVec` can be added pretty fast.

**Comment by Amanieu** [comment]
OK, if you feel that it won't take long then we can do the crater run with all collections. It'll take a few days until we reach the top of the queue anyways.

**Comment by TimDiekmann** [comment]
Is it possible to simply update the commit used by crater?

**Comment by Amanieu** [comment]
@craterbot end=ac6afa7e002222176f3dcb43f18522047891b080

**Comment by TimDiekmann** [comment]
I think the most important ones are implemented: `Box`, `Vec`, and `String`. In addition to `VecDeque` and `BinaryHeap` (which where straight forward due to `(Raw)Vec` I may add `Rc` and `Arc`. As the trees are pretty complicated I leave them out for now and maybe I add an implementation for `CString` and `OsString`.

**Comment by Amanieu** [comment]
@craterbot end=d93021f5faa26e4f9d2719e79f9974e06aa40e17

**Comment by TimDiekmann** [comment]
For the latest commit a new try-run is not required. The crater run can still be used with d93021f.

**Comment by Mark-Simulacrum** [comment]
Is there a reason we're not updating the start= commit of the crater run? That'll make it contain any regressions that happened since then on master too which seems unhelpful?

**Comment by Amanieu** [comment]
@Mark-Simulacrum That's an oversight, I'm not too familiar with craterbot's API.

@craterbot start=7c59a81a5fcbaaca311f744cd7c68d99bfbb05d3

**Comment by TimDiekmann** [comment]
Rebased, but I think I need some help with the chalk-error.

**Comment by pietroalbini** [comment]
@craterbot start=master#7c59a81a5fcbaaca311f744cd7c68d99bfbb05d3 end=try#d93021f5faa26e4f9d2719e79f9974e06aa40e17

The branch name still needs to be added.

**Comment by pietroalbini** [comment]
@craterbot start=master#7c59a81a5fcbaaca311f744cd7c68d99bfbb05d3 end=try#d93021f5faa26e4f9d2719e79f9974e06aa40e17

**Comment by jackh726** [comment]
@nikomatsakis ping about the chalk error

```
thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `2`,
 right: `1`: Only fundamental types with a single parameter are supported'
```

**Comment by pietroalbini** [comment]
The "unknown" are a Crater bug, we had so many Crater bugs with this run :(

**Comment by nikomatsakis** [comment]
The chalk error is legit in the sense that we haven't defined what fundamental means for a type with multiple type parameters. I guess we will ignore all but the first, but we're going to have to spend at least some time thinking about how to use this to break coherence :)

**Comment by TimDiekmann** [comment]
I finally had time to get through the crater report. There are 100 regressed crates, and everyone failed because `Box::alloc` and `Box::alloc_mut`. Previously, they `Deref`ed to `T` but now `Box::alloc(&self)` is called instead. As long as we add methods to structs, which implements `Deref`, which takes `Self` instead of `&self` we should be compatible.

As this is still blocked on #72314 and the chalk error, I'll close this. Actually, this PR was never designed to be merged anyway 😉

**Pull request closed without merging**